### PR TITLE
Update deploy

### PR DESCRIPTION
**Pull request by jliu70**
Update deploy to detect for MacOS (Darwin) platform per #1

### CODE REVIEW

File: deploy [hunk 1]
```diff
@@ -24,12 +24,23 @@
...
 import argparse
 import glob
 import json
 import re
 from deploy_config import config
 
+# Detect OS - set "-X" option for MacOS (Darwin)
+p = platform.system()
+if p == "Darwin":
+    print(f"platform system is {p} - setting MacOS option to cp commands")
```
Comment (rwhite-lightbox) [inline deploy:37]: print the platform system once and skip the mac specifics...

File: deploy [hunk 2]
```diff
@@ -83,7 +94,11 @@ cp_lib_dir = os.path.join(args.circuitpy, "lib")
 
 def copy_lib(lib_name):
     print("  " + lib_name)
-    os.system("cp -R {0} {1}".format(os.path.join(lib_dir, lib_name), cp_lib_dir))
+    if cpswitch == True:
+        print(f"MacOS detected - adding -X switch for cp")
```
Comment (rwhite-lightbox) [inline deploy]: redundant, see above

File: deploy [hunk 3]
```diff
@@ -117,7 +132,11 @@ def replace_file(fname):
             print("  " + fname)
             if os.path.exists(dest_path):
                 os.system("rm -r {0}".format(dest_path))
-            os.system("cp -R {0} {1}".format(fname, args.circuitpy))
+            if cpswitch == True:
+                print(f"MacOS detected - adding -X switch for cp")
```
Comment (rwhite-lightbox) [inline deploy]: redundant, see above...

**Review by rwhite-lightbox — CHANGES_REQUESTED** [review]
Looks good and aligns with what I was able to get working.  Do we also need a "cp -X" on line 171?

File: deploy (hunk 2 above)
Comment (jliu70) [reply deploy]: Yes, this makes sense.  Should only print once at the beginning.  These print statements on lines 98 and 136 are extraneous.

**Comment by jliu70** [comment]
I didn't add the switch on line 171 since it is copying a single file
`code.py`, but I suppose we should add it there too just in case.

Thanks!

**Comment by jliu70** [comment]
@RWhite-EDR  So I created a windows version.  Basically it will use the xcopy command.
For example:
```
    elif cpswitch == "Windows":
        os.system("xcopy /S /Y /F {0} {1}".format(fname, args.circuitpy))
    else:
```

Two issues with Windows:
1. The shebang for Windows will conflict.  Usually Python3 is installed on Windows as a separate package and depending on the path installed (and possibly if a virtual env is being used), it would likely be better for Windows to use `#!/usr/bin/env python`.  For Powershell and command prompt, you will still need to run via `python deploy-windows` as these will not honor the unix-style shebang.  For Git Bash for Windows it does honor the the shebang, but best to use `#!/usr/bin/env python`.  The shebang of `#!/usr/bin/python3` will not matter for Powershell or command prompt.
2. The directory separator for deploy_config.py or deploy_config.json will need to be escaped with double "\\", and the unix-style "/" separator will not work.  This is probably just something to update in the documentation.

I haven't pushed the code up, but I suppose I can do that if you want to see it.  Maybe push a new file `deploy-windows` for now?

**Review by dastels — APPROVED** [review]

**Comment by dastels** [comment]
Thanks for the improvements! I only work on Ubuntu so support for other platforms is welcomed.

Dave

**Comment by rwhite-lightbox** [comment]
@dastels thanks for the commit and great script.
@jliu70 Understood on windows and creating a separate deploy-windows makes sense for now.  I need to check my notes, but I wasn't able to get this to work with xcopy.  I was probably missing a switch.

I don't think having a requirement to run using git bash is a bad idea.  Are you suggesting that switching the shebang to #!/usr/bin/env python for all three OS is the way to go or just for Windows?

Great job!

**Comment by jliu70** [comment]
@RWhite-EDR  I was able to get it working with xcopy.  The code snippet above is basically the xcopy command that I used.   You may give it a try and it should work.   

With regards to Windows and the shebang - this is mostly due to differences with how Python3 is installed on Windows compared to Linux or MacOS.    The shebang only affects Git bash for Windows, as Powershell and command prompt ignores the shebang.   If we add the above snippets to the deploy script, it should work for Windows powershell and command prompt just fine as long as the double "\\" is used within the config file(s) for the paths.
For Git Bash for Windows - there's no easy way to address this other editing the shebang to be `#!/usr/bin/env python` or commenting the shebang out.    

So to summarize - the issue is with Git Bash for Windows.   Possibly the best approach is to document that for Git Bash for Windows to comment out the first line or to modify it to `#!/usr/bin/env python`.

**Comment by jliu70** [comment]
@RWhite-EDR   I pushed a new branch updateDeployWindows which you can check out.

**Comment by rwhite-lightbox** [comment]
@jliu70 Sorry for the delay.  I agree that documenting around the git bash issue is probably the simplest.

For the code changes, rather than test for cpswitch all the way down, could we define two variables at the top based on the OS and then format in both the copy and remove commands?

if (platform.system =- 'Darwin'):
     cpCmd = 'cp -RX'
elif (platform.system == 'Windows'):
    cpCmd = 'xcopy /S /Y /F'
else:
    cpCmd = 'cp -R'

os.system("{0} {1} {2}".format(cpCmd, os.path.join(lib_dir, lib_name), cp_lib_dir))

**Comment by dastels** [comment]
Definitely a better approach. Please write, test, and submit a PR.

**Comment by jliu70** [comment]
Okay, I have updated the code and submit a PR.

https://github.com/dastels/circuitpython_deploy/pull/8
